src/test/java/org/wildfly/security/authz/RoleMappingTest.java

```diff
@@ -0,0 +1,253 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.authz;
+
+import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Basic String role mapping tests
+ *
+ * @author <a href="mailto:mmazanek@redhat.com">Martin Mazanek</a>
+ */
+public class RoleMappingTest {
+
+    @Test
+    public void testMappedRoles1() {
+        Roles roles = createRoles("foo", "joe");
+
+        Map<String, Set<String>> mappingMap = new HashMap<>();
+        mappingMap.put("foo", createSet("bar", "role"));
+
+        RoleMapper mapper = new MappedRoleMapper.Builder()
+                .setRemoveMapped(true)
+                .setRemoveNonMapped(true)
+                .setRoleMap(mappingMap).build();
+
+        Roles mappedRoles = mapper.mapRoles(roles);
+
+        assertTrue(mappedRoles.contains("bar"));
+        assertTrue(mappedRoles.contains("role"));
+        assertFalse(mappedRoles.contains("foo"));
+        assertFalse(mappedRoles.contains("joe"));
+
+        Iterator<String> iterator = mappedRoles.iterator();
+        int count = 0;
+        while (iterator.hasNext()) {
+            iterator.next();
+            count++;
+        }
+        assertTrue(count == 2);
+    }
+
+    @Test
+    public void testMappedRoles2() {
+        Roles roles = createRoles("foo", "joe");
+
+        Map<String, Set<String>> mappingMap = new HashMap<>();
+        mappingMap.put("foo", createSet("bar", "role"));
+
+        RoleMapper mapper = new MappedRoleMapper.Builder()
+                .setRemoveMapped(false)
+                .setRemoveNonMapped(true)
+                .setRoleMap(mappingMap).build();
+
+        Roles mappedRoles = mapper.mapRoles(roles);
+
+        assertTrue(mappedRoles.contains("bar"));
+        assertTrue(mappedRoles.contains("role"));
+        assertTrue(mappedRoles.contains("foo"));
+        assertFalse(mappedRoles.contains("joe"));
+
+        Iterator<String> iterator = mappedRoles.iterator();
+        int count = 0;
+        while (iterator.hasNext()) {
+            iterator.next();
+            count++;
+        }
+        assertTrue(count == 3);
+    }
+
+
+    @Test
+    public void testMappedRoles3() {
+        Roles roles = createRoles("foo", "joe");
+
+        Map<String, Set<String>> mappingMap = new HashMap<>();
+        mappingMap.put("foo", createSet("bar", "role"));
+
+        RoleMapper mapper = new MappedRoleMapper.Builder()
+                .setRemoveMapped(true)
+                .setRemoveNonMapped(false)
```

Comment:
Could you add tests when mapping is not successful at all?

Reply:
Yes.
I meant when user role(s) are not among Map keys. And setRemoveNonMapped is false/true.
Or when mappings are configured to deny access like role1 -> ""
Or when user has no roles (empty or null)